should this error be severe? or an error?

I edited my appengine-web.xml file so that it was malformed and saved (not uncommon during development). This stack trace showed up in the console. I'm not sure this should be a severe error. 

For comparison, a syntax error in a Java file does not log a similar exception message. 

Eclipse does put up a problem for this in the problems view. I suggest we leave it at that and not log here.

```
Apr 29, 2017 6:44:38 AM com.google.cloud.tools.eclipse.appengine.standard.java8.AppEngineWebBuilder checkRuntimeElement
SEVERE: P/fred: error updating facets
java.io.IOException: Cannot parse appengine-web.xml
	at com.google.cloud.tools.appengine.AppEngineDescriptor.parse(AppEngineDescriptor.java:55)
	at com.google.cloud.tools.eclipse.appengine.standard.java8.AppEngineWebBuilder.checkRuntimeElement(AppEngineWebBuilder.java:82)
	at com.google.cloud.tools.eclipse.appengine.standard.java8.AppEngineWebBuilder.build(AppEngineWebBuilder.java:71)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:734)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:205)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:245)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:300)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:303)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:359)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:382)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:144)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:235)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: org.xml.sax.SAXParseException; lineNumber: 5; columnNumber: 19; The element type "runtime" must be terminated by the matching end-tag "</runtime>".
	at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:257)
	at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:347)
	at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:121)
	at com.google.cloud.tools.appengine.AppEngineDescriptor.parse(AppEngineDescriptor.java:53)
	... 14 more
```

--- a/plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/META-INF/MANIFEST.MF
@@ -8,7 +8,7 @@ Bundle-Vendor: %providerName
 Bundle-Localization: plugin
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-ActivationPolicy: lazy
-Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.2",
+Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.3",
  org.eclipse.ui.workbench,
  org.eclipse.core.commands
 Export-Package: com.google.cloud.tools.eclipse.appengine.deploy.ui.flexible

--- a/plugins/com.google.cloud.tools.eclipse.appengine.facets/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.facets/META-INF/MANIFEST.MF
@@ -7,7 +7,7 @@ Bundle-Version: 0.1.0.qualifier
 Bundle-Vendor: Google Inc.
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-Localization: plugin
-Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.2",
+Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.3",
  com.google.cloud.tools.eclipse.sdk,
  com.google.guava;bundle-version="[20.0.0,21.0.0)",
  org.eclipse.core.resources,

--- a/plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ModuleUtilsTest.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ModuleUtilsTest.java
@@ -49,7 +49,7 @@ public class ModuleUtilsTest {
     Logger logger = Logger.getLogger(ModuleUtils.class.getName());
     logger.setLevel(Level.OFF);
   }
-  
+
   @Before
   public void setUpMocks() {
     IProject project = Mockito.mock(IProject.class);
@@ -67,34 +67,40 @@ public class ModuleUtilsTest {
   public void testGetServiceId_null() {
     ModuleUtils.getServiceId(null);
   }
-  
+
   @Test
   public void testGetServiceId_noWebInf() {
     Mockito.when(webinf.exists()).thenReturn(false);
     Assert.assertEquals("default", ModuleUtils.getServiceId(module));
   }
-  
+
   @Test
   public void testGetServiceId_default() throws CoreException {
     Mockito.when(descriptorFile.getContents()).thenThrow(new CoreException(Status.CANCEL_STATUS));
     Assert.assertEquals("default", ModuleUtils.getServiceId(module));
   }
-  
+
   @Test
   public void testGetServiceId() throws CoreException {
     mockAppEngineWebXml("appengine-web.xml");
     Assert.assertEquals("myServiceId", ModuleUtils.getServiceId(module));
   }
- 
+
   @Test
   public void testGetServiceId_module() throws CoreException {
     mockAppEngineWebXml("appengine-web_module.xml");
     Assert.assertEquals("myServiceId", ModuleUtils.getServiceId(module));
   }
-  
+
   @Test
   public void testGetServiceId_notPresent() throws CoreException {
     mockAppEngineWebXml("appengine-web_noservice.xml");
+    Assert.assertEquals("default", ModuleUtils.getServiceId(module));
+  }
+
+  @Test
+  public void testGetServiceId_malformedXml() throws CoreException {
+    mockAppEngineWebXml("appengine-web_malformed.xml");
     Assert.assertEquals("default", ModuleUtils.getServiceId(module));
   }
 

--- a/plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/appengine-web_malformed.xml
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/appengine-web_malformed.xml
@@ -1,0 +1,1 @@
+<malformed>

--- a/plugins/com.google.cloud.tools.eclipse.appengine.localserver/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.localserver/META-INF/MANIFEST.MF
@@ -9,7 +9,7 @@ Bundle-Localization: plugin
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-Activator: com.google.cloud.tools.eclipse.appengine.localserver.Activator
 Bundle-ActivationPolicy: lazy
-Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.2",
+Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.3",
  com.google.cloud.tools.eclipse.appengine.ui;bundle-version="0.1.0",
  com.google.guava;bundle-version="[20.0.0,21.0.0)",
  javax.servlet;bundle-version="2.5.0",

--- a/plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ModuleUtils.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ModuleUtils.java
@@ -30,6 +30,7 @@ import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.Path;
 import org.eclipse.wst.server.core.IModule;
 import org.eclipse.wst.server.core.IServer;
+import org.xml.sax.SAXException;
 
 /**
  * A set of utility methods for dealing with WTP {@link IModule}s.
@@ -40,7 +41,7 @@ public class ModuleUtils {
   /**
    * Retrieve the &lt;service&gt; or &lt;module&gt; identifier from <tt>appengine-web.xml</tt>.
    * If an identifier is not found, return "default".
-   * 
+   *
    * @return the identifier, defaulting to "default" if not found
    */
   public static String getServiceId(IModule module) {
@@ -53,11 +54,11 @@ public class ModuleUtils {
         if (serviceId != null) {
           return serviceId;
         }
-      } catch (CoreException | IOException ex) {
+      } catch (CoreException | IOException | SAXException ex) {
         logger.log(Level.WARNING, "Unable to read " + descriptorFile.getFullPath(), ex);
       }
     }
-    
+
     return "default";
   }
 

--- a/plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ModuleUtils.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ModuleUtils.java
@@ -54,7 +54,9 @@ public class ModuleUtils {
         if (serviceId != null) {
           return serviceId;
         }
-      } catch (CoreException | IOException | SAXException ex) {
+      } catch (SAXException ex) {
+        // Parsing failed due to malformed XML; return "default".
+      } catch (CoreException | IOException ex) {
         logger.log(Level.WARNING, "Unable to read " + descriptorFile.getFullPath(), ex);
       }
     }

--- a/plugins/com.google.cloud.tools.eclipse.appengine.newproject.maven/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.newproject.maven/META-INF/MANIFEST.MF
@@ -7,7 +7,7 @@ Bundle-Version: 0.1.0.qualifier
 Bundle-Vendor: Google Inc.
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-Localization: plugin
-Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.2",
+Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.3",
  com.google.cloud.tools.eclipse.appengine.newproject;bundle-version="0.1.0",
  com.google.cloud.tools.eclipse.appengine.ui,
  com.google.guava;bundle-version="[20.0.0,21.0.0)",

--- a/plugins/com.google.cloud.tools.eclipse.appengine.newproject/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.newproject/META-INF/MANIFEST.MF
@@ -8,7 +8,7 @@ Bundle-Vendor: Google Inc.
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-Localization: plugin
 Bundle-ActivationPolicy: lazy
-Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.2",
+Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.3",
  com.google.cloud.tools.eclipse.appengine.facets,
  com.google.cloud.tools.eclipse.appengine.ui,
  javax.servlet;bundle-version="2.5.0",

--- a/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8.test/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8.test/META-INF/MANIFEST.MF
@@ -6,7 +6,7 @@ Fragment-Host: com.google.cloud.tools.eclipse.appengine.standard.java8
 Bundle-Version: 0.1.0.qualifier
 Bundle-Vendor: Google, Inc.
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
-Import-Package: com.google.cloud.tools.appengine;version="0.3.2",
+Import-Package: com.google.cloud.tools.appengine;version="0.3.3",
  com.google.cloud.tools.eclipse.appengine.facets.convert,
  com.google.cloud.tools.eclipse.test.util.project,
  org.eclipse.jst.j2ee.web.project.facet,

--- a/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8.test/src/com/google/cloud/tools/eclipse/appengine/standard/java8/ConversionTests.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8.test/src/com/google/cloud/tools/eclipse/appengine/standard/java8/ConversionTests.java
@@ -34,6 +34,7 @@ import org.eclipse.jst.j2ee.web.project.facet.WebFacetUtils;
 import org.eclipse.wst.common.project.facet.core.IFacetedProject;
 import org.junit.Rule;
 import org.junit.Test;
+import org.xml.sax.SAXException;
 
 /**
  * Test that our <em>Convert to App Engine Standard Project</em> doesn't downgrade Java or jst.web
@@ -45,7 +46,8 @@ public class ConversionTests {
       new TestProjectCreator().withFacetVersions(JavaFacet.VERSION_1_8, WebFacetUtils.WEB_31);
 
   @Test
-  public void testChangeToAndFrom_1_8() throws CoreException, IOException, InterruptedException {
+  public void testChangeToAndFrom_1_8()
+      throws CoreException, IOException, InterruptedException, SAXException {
     IFacetedProject project = testProject.getFacetedProject();
     Job conversionJob = new AppEngineStandardProjectConvertJob(project);
     conversionJob.schedule();

--- a/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8.test/src/com/google/cloud/tools/eclipse/appengine/standard/java8/JavaFacetChangeTests.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8.test/src/com/google/cloud/tools/eclipse/appengine/standard/java8/JavaFacetChangeTests.java
@@ -37,6 +37,7 @@ import org.eclipse.wst.common.project.facet.core.IFacetedProject;
 import org.eclipse.wst.common.project.facet.core.IFacetedProject.Action;
 import org.junit.Rule;
 import org.junit.Test;
+import org.xml.sax.SAXException;
 
 /**
  * Test changing the Java Facet to/from Java 8 results in {@code <runtime>} changes in the
@@ -47,9 +48,9 @@ public class JavaFacetChangeTests {
   public TestProjectCreator testProject = new TestProjectCreator()
       .withFacetVersions(JavaFacet.VERSION_1_7, WebFacetUtils.WEB_31,
           AppEngineStandardFacet.FACET_VERSION);
-  
+
   @Test
-  public void testChangeToAndFrom_1_8() throws CoreException, IOException {
+  public void testChangeToAndFrom_1_8() throws CoreException, IOException, SAXException {
     IFacetedProject project = testProject.getFacetedProject();
     IFile descriptorFile =
         WebProjectUtil.findInWebInf(project.getProject(), new Path("appengine-web.xml"));
@@ -64,7 +65,7 @@ public class JavaFacetChangeTests {
     actions.add(new Action(Action.Type.VERSION_CHANGE, JavaFacet.VERSION_1_8, null));
     project.modify(actions, null);
     ProjectUtils.waitForProjects(project.getProject());
-    
+
     try (InputStream is = descriptorFile.getContents()) {
       AppEngineDescriptor descriptor = AppEngineDescriptor.parse(is);
       assertTrue(descriptor.isJava8());

--- a/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/META-INF/MANIFEST.MF
@@ -9,7 +9,7 @@ Require-Bundle: com.google.cloud.tools.eclipse.appengine.localserver;bundle-vers
  org.eclipse.wst.common.project.facet.core,
  org.eclipse.core.resources,
  org.eclipse.core.runtime
-Import-Package: com.google.cloud.tools.appengine;version="0.3.2",
+Import-Package: com.google.cloud.tools.appengine;version="0.3.3",
  com.google.cloud.tools.eclipse.appengine.facets,
  com.google.cloud.tools.eclipse.util,
  com.google.common.io;version="[20.0.0,21.0.0)",

--- a/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/src/com/google/cloud/tools/eclipse/appengine/standard/java8/AppEngineWebBuilder.java
+++ b/plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/src/com/google/cloud/tools/eclipse/appengine/standard/java8/AppEngineWebBuilder.java
@@ -38,6 +38,7 @@ import org.eclipse.jst.j2ee.web.project.facet.WebFacetUtils;
 import org.eclipse.wst.common.project.facet.core.IFacetedProject;
 import org.eclipse.wst.common.project.facet.core.IFacetedProject.Action;
 import org.eclipse.wst.common.project.facet.core.ProjectFacetsManager;
+import org.xml.sax.SAXException;
 
 /**
  * Builder to monitor user changes relating to the {@code <runtime>java8</runtime>} element in
@@ -83,7 +84,7 @@ public class AppEngineWebBuilder extends IncrementalProjectBuilder {
       boolean hasJava8Facet = project.hasProjectFacet(JavaFacet.VERSION_1_8);
       // if not the same, then we update the facet to match the appengine-web.xml
       if (hasJava8Facet != hasJava8Runtime) {
-        Set<Action> updates = new HashSet<Action>();
+        Set<Action> updates = new HashSet<>();
         // Can upgrade jst.web to 3.1, but cannot downgrade from 3.1
         if (hasJava8Runtime) {
           updates.add(new Action(Action.Type.VERSION_CHANGE, JavaFacet.VERSION_1_8, null));
@@ -94,6 +95,8 @@ public class AppEngineWebBuilder extends IncrementalProjectBuilder {
         logger.fine(getProject() + ": changing facets: " + updates);
         project.modify(updates, monitor);
       }
+    } catch (SAXException ex) {
+      // Parsing failed due to malformed XML; just don't check the value now.
     } catch (CoreException | IOException ex) {
       logger.log(Level.SEVERE, getProject() + ": error updating facets", ex);
     }

--- a/plugins/com.google.cloud.tools.eclipse.sdk.ui/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.sdk.ui/META-INF/MANIFEST.MF
@@ -9,7 +9,7 @@ Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-ActivationPolicy: lazy
 Bundle-Localization: plugin
 Require-Bundle: org.eclipse.ui;bundle-version="3.107.0",
- com.google.cloud.tools.appengine;bundle-version="0.3.2",
+ com.google.cloud.tools.appengine;bundle-version="0.3.3",
  org.eclipse.equinox.preferences;bundle-version="3.5.300",
  org.eclipse.equinox.common;bundle-version="3.7.0",
  org.eclipse.e4.core.contexts;bundle-version="1.4.0",

--- a/plugins/com.google.cloud.tools.eclipse.sdk/META-INF/MANIFEST.MF
+++ b/plugins/com.google.cloud.tools.eclipse.sdk/META-INF/MANIFEST.MF
@@ -7,7 +7,7 @@ Bundle-Version: 0.1.0.qualifier
 Bundle-Vendor: Google Inc.
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-ActivationPolicy: lazy
-Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.2",
+Require-Bundle: com.google.cloud.tools.appengine;bundle-version="0.3.3",
  org.eclipse.equinox.preferences;bundle-version="3.5.300",
  org.eclipse.e4.core.contexts;bundle-version="1.4.0",
  org.eclipse.e4.core.di;bundle-version="1.5.0"

--- a/pom.xml
+++ b/pom.xml
@@ -147,7 +147,7 @@
       <!-- bundle: com.google.cloud.tools.appengine -->
       <groupId>com.google.cloud.tools</groupId>
       <artifactId>appengine-plugins-core</artifactId>
-      <version>0.3.2</version>
+      <version>0.3.3</version>
     </dependency>
     <dependency>
       <!-- bundle: com.google.guava -->

